core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java

```diff
@@ -0,0 +1,181 @@
+
+    private static final int ACTIVE_SHARD_COUNT_DEFAULT = -2;
+    private static final int ALL_ACTIVE_SHARDS = -1;
+
+    public static final ActiveShardCount DEFAULT = new ActiveShardCount(ACTIVE_SHARD_COUNT_DEFAULT);
+    public static final ActiveShardCount ALL = new ActiveShardCount(ALL_ACTIVE_SHARDS);
+    public static final ActiveShardCount NONE = new ActiveShardCount(0);
+    public static final ActiveShardCount ONE = new ActiveShardCount(1);
+
+    private final int value;
+
+    private ActiveShardCount(final int value) {
+        this.value = value;
+    }
+
+    /**
+     * Get an ActiveShardCount instance for the given value.  The value is first validated to ensure
+     * it is a valid shard count and throws an IllegalArgumentException if validation fails.  Valid
+     * values are any non-negative number.  Directly use {@link ActiveShardCount#DEFAULT} for the
+     * default value (which is half the shards) or {@link ActiveShardCount#ALL} to specify all the shards.
+     */
+    public static ActiveShardCount from(final int value) {
+        if (value < 0) {
+            throw new IllegalArgumentException("shard count cannot be a negative value");
+        }
+        return get(value);
+    }
+
+    private static ActiveShardCount get(final int value) {
+        switch (validateValue(value)) {
+            case ACTIVE_SHARD_COUNT_DEFAULT:
+                return DEFAULT;
+            case ALL_ACTIVE_SHARDS:
+                return ALL;
+            case 0:
+                return NONE;
+            case 1:
+                return ONE;
+            default:
+                return new ActiveShardCount(value);
+        }
+    }
+
+    @Override
+    public void writeTo(final StreamOutput out) throws IOException {
+        out.writeInt(value);
+    }
+
+    public static ActiveShardCount readFrom(final StreamInput in) throws IOException {
+        return get(in.readInt());
+    }
+
+    private static int validateValue(final int value) {
+        if (value < 0 && value != ACTIVE_SHARD_COUNT_DEFAULT && value != ALL_ACTIVE_SHARDS) {
+            throw new IllegalArgumentException("Invalid ActiveShardCount[" + value + "]");
+        }
+        return value;
+    }
+
+    /**
+     * Resolve this instance to an actual integer value for the number of active shard counts.
+     * If {@link ActiveShardCount#DEFAULT} or {@link ActiveShardCount#ALL} are specified, then
+     * the given {@link IndexShardRoutingTable} is used to determine the actual active shard count.
+     * The default value indicates half the shards must be active, but if there is only one shard
+     * in the index, then the value for the default will be 1 (instead of 0).
+     */
+    public int resolve(final IndexShardRoutingTable shardRoutingTable) {
+        Objects.requireNonNull(shardRoutingTable);
+        if (this == ActiveShardCount.DEFAULT) {
+            int val = shardRoutingTable.size() / 2; // default is half the shards must be active
```

Comment:
6 shard copies? That's rather useless in a system like ES. Reasonable choices are 1, 2 or 3 shard copies.

For 1 copy, it trivially makes sense to choose wait_for_active_shards = 1 as default. Fixing defaults for 2 or 3 shard copies feels highly arbitrary to me, there is no good default for all situations. If we always pick wait_for_active_shards = 1 as default, it is easy to document, simple to understand and does not imply any kind of quorum semantics.

This setting is of limited use anyhow as it does not provide the guarantee that most users are after, namely that a write has made it to a certain number of shard copies once we acknowledge the write.  Maybe the better approach would be to remove this setting for now (at the replication operation level, not index creation) until we've implemented one that gives guarantees on the number of shard copies that have received a write.